File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/wallet/components/IdenticonViewerSheet.kt
package app.getvela.wallet.feature.wallet.components

import app.getvela.wallet.core.platform.Clipboard
import android.content.Context
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.times
import app.getvela.wallet.core.designsystem.components.VelaPrimaryButton
import app.getvela.wallet.core.designsystem.components.VelaSecondaryButton
import app.getvela.wallet.core.designsystem.theme.VelaTheme
import app.getvela.wallet.core.designsystem.tokens.VelaFontFamily
import app.getvela.wallet.core.designsystem.tokens.VelaFontWeight
import app.getvela.wallet.core.designsystem.tokens.VelaLeading
import app.getvela.wallet.core.designsystem.tokens.VelaRadius
import app.getvela.wallet.core.designsystem.tokens.VelaSizing
import app.getvela.wallet.core.designsystem.tokens.VelaSpacing
import app.getvela.wallet.core.designsystem.tokens.VelaTextSize
import app.getvela.wallet.core.i18n.I18nKeys
import app.getvela.wallet.core.i18n.LocalVelaStrings

/** How long the copy confirmation holds before the label reverts. */
private const val COPIED_FEEDBACK_MS = 1_500L

/**
 * The identicon, big, above the address that drew it (founder call,
 * 2026-08-26).
 *
 * The artwork is a fingerprint of the address: the same address always draws
 * the same pattern, which is only useful once somebody has seen the two
 * together often enough to recognise one from the other. A 40dp avatar in a
 * header never teaches that. This does — and it opens from the artwork itself,
 * wherever the artwork is drawn, rather than from a settings page nobody
 * visits.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun IdenticonViewerSheet(
    /** The seed, verbatim: what the artwork was drawn from. */
    address: String,
    onDismiss: () -> Unit,
    /** Spec 049: the name the opener showed, so the initials style draws the same disc big. */
    name: String? = null,
) {
    val strings = LocalVelaStrings.current
    val colors = VelaTheme.colors
    val context = LocalContext.current
    var copied by remember { mutableStateOf(false) }

    LaunchedEffect(copied) {
        if (copied) {
            kotlinx.coroutines.delay(COPIED_FEEDBACK_MS)
            copied = false
        }
    }

    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = colors.bgRaised) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(
                    start = VelaSizing.screenPaddingX,
                    end = VelaSizing.screenPaddingX,
                    bottom = VelaSpacing.xl4,
                ),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(VelaSpacing.lg),
        ) {
            IdenticonAvatar(
                tappable = false,
                seed = address,
                size = WalletMetrics.identiconViewerSize,
                contentDescription = null,
                name = name,
            )

            Text(
                text = strings.t(I18nKeys.Wallet.IDENTICON_TITLE),
                color = colors.fgBase,
                fontFamily = VelaFontFamily,
                fontWeight = VelaFontWeight.bold,
                fontSize = VelaTextSize.xl,
            )

            Text(
                text = strings.t(I18nKeys.Wallet.IDENTICON_CAPTION),
                color = colors.fgMuted,
                fontFamily = VelaFontFamily,
                fontWeight = VelaFontWeight.regular,
                fontSize = VelaTextSize.base,
                lineHeight = VelaLeading.relaxed * VelaTextSize.base,
                textAlign = TextAlign.Center,
            )

            // The WHOLE address, wrapped rather than middle-truncated: a
            // fingerprint you can only see half of teaches half a habit.
            Text(
                text = address,
                color = colors.fgBase,
                fontFamily = VelaFontFamily,
                fontWeight = VelaFontWeight.regular,
                fontSize = VelaTextSize.sm,
                lineHeight = VelaLeading.relaxed * VelaTextSize.sm,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .fillMaxWidth()
                    .background(colors.bgSunken, RoundedCornerShape(VelaRadius.lg))
                    .padding(VelaSpacing.lg),
            )

            VelaPrimaryButton(
                text = strings.t(
                    if (copied) I18nKeys.Wallet.IDENTICON_COPIED else I18nKeys.Wallet.COPY_ADDRESS,
                ),
                onClick = {
                    copyToClipboard(context, address)
                    copied = true
                },
                modifier = Modifier.fillMaxWidth(),
            )
            VelaSecondaryButton(
                text = strings.t(I18nKeys.Wallet.IDENTICON_CLOSE),
                onClick = onDismiss,
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}

private fun copyToClipboard(context: Context, address: String): Boolean =
    Clipboard.copy(context, "address", address)

File: app-ios/VelaWallet/VelaWallet/Components/Wallet/IdenticonViewerSheet.swift
//
//  IdenticonViewerSheet.swift
//  VelaWallet
//
//  The identicon, big, above the address that drew it (founder call,
//  2026-08-26).
//
//  The artwork is a fingerprint of the address: the same address always draws
//  the same pattern, which is only useful once somebody has seen the two
//  together often enough to recognise one from the other. A 40pt avatar in a
//  header never teaches that. This does — and it opens from the artwork itself,
//  wherever the artwork is drawn, rather than from a settings page nobody
//  visits.
//

import SwiftUI

struct IdenticonViewerSheet: View {
    @Environment(\.theme) private var theme
    let loc: Loc
    /// The seed, verbatim: what the artwork was drawn from.
    let address: String
    /// Whose it is, where the caller knows — the initials style draws from the
    /// NAME, so a viewer without one would show a different face than the row
    /// it was opened from.
    var name: String?
    let onClose: () -> Void

    @State private var copied = false

    var body: some View {
        // SCROLLS, and pads past the drag indicator.
        //
        // The content — a big circle, a title, a wrapped 42-character address
        // and two buttons — is taller than the `.medium` detent it opens at,
        // and a bare `VStack` in a sheet does not scroll: it clips. On the
        // device that read as artwork sliced by the sheet's top edge with the
        // system grabber sitting on it, and a 关闭 button cut off below the
        // screen — the way out of the sheet, unreachable (founder,
        // 2026-09-16). Scrolling makes every element reachable at ANY detent,
        // which is the property that has to hold however tall the copy runs in
        // a language nobody has translated yet.
        ScrollView {
        VStack(spacing: Tokens.Space.s16) {
            // Not tappable: it is already the viewer.
            IdenticonAvatar(seed: address, size: WalletGeometry.identiconViewer,
                            name: name, tappable: false)
                .padding(.bottom, Tokens.Space.s8)

            Text(loc.t("componentsUi.identiconViewer.title"))
                .typeRole(Typography.title)
                .foregroundStyle(theme.fgBase)

            Text(loc.t("componentsUi.identiconViewer.caption"))
                .typeRole(Typography.body)
                .foregroundStyle(theme.fgMuted)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)

            // The WHOLE address, wrapped rather than middle-truncated: a
            // fingerprint you can only see half of teaches half a habit.
            Text(verbatim: address)
                .monoRole(Typography.monoAddress)
                .foregroundStyle(theme.fgBase)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)
                .frame(maxWidth: .infinity)
                .padding(Tokens.Space.s16)
                .background(
                    RoundedRectangle(cornerRadius: Tokens.Radius.r12)
                        .fill(theme.bgSunken)
                )

            VStack(spacing: Tokens.Space.s12) {
                VelaButton(
                    title: loc.t(copied
                        ? "componentsUi.identiconViewer.copied"
                        : "componentsUi.identiconViewer.copyAddress"),
                    kind: .primary,
                    action: copy
                )
                VelaButton(
                    title: loc.t("componentsUi.identiconViewer.close"),
                    kind: .secondary,
                    action: onClose
                )
            }
            .padding(.top, Tokens.Space.s8)
        }
        .padding(.horizontal, Tokens.Layout.screenPaddingX)
        // Top clears the system drag indicator, which is drawn OVER the
        // content rather than above it.
        .padding(.top, Tokens.Space.s32)
        .padding(.bottom, Tokens.Space.s32)
        .frame(maxWidth: .infinity)
        }
        .scrollBounceBehavior(.basedOnSize)
        .presentationDragIndicator(.visible)
        .presentationBackground(theme.bgRaised)
    }

    private func copy() {
        #if canImport(UIKit)
        UIPasteboard.general.string = address
        #endif
        copied = true
        Task {
            try? await Task.sleep(for: .seconds(Interaction.copiedFeedbackSeconds))
            copied = false
        }
    }
}

#Preview("Identicon viewer") {
    IdenticonViewerSheet(
        loc: Loc(overrideTag: "zh"),
        address: "0x14fB1fB21751E29F7Ec48dC450017552E3D1eA5c",
        onClose: {}
    )
    .themed(.light)
    .environment(\.identiconProvider, .previewSafe)
}
